purchase_triple_discount/models/purchase_order_line.py

```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Tecnativa - David Vidal
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+from odoo.addons import decimal_precision as dp
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = "purchase.order.line"
+
+    discount2 = fields.Float(
+        'Discount 2 (%)',
+        digits=dp.get_precision('Discount'),
+        default=0.0,
+    )
+    discount3 = fields.Float(
+        'Discount 3 (%)',
+        digits=dp.get_precision('Discount'),
+        default=0.0,
+    )
+
+    _sql_constraints = [
+        ('discount2_limit', 'CHECK (discount2 <= 100.0)',
+         'Discount 2 must be lower than 100%.'),
+        ('discount3_limit', 'CHECK (discount3 <= 100.0)',
+         'Discount 3 must be lower than 100%.'),
+    ]
+
+    @api.depends('product_qty', 'price_unit', 'taxes_id', 'discount',
+                 'discount2', 'discount3')
+    def _compute_amount(self):
+        self._triple_discount(super(PurchaseOrderLine, self)._compute_amount)
+
+    def _get_stock_move_price_unit(self):
+        self._triple_discount(
+            super(PurchaseOrderLine, self)._get_stock_move_price_unit)
+
+    def _triple_discount(self, call_super):
+        """
+        Apply three succesive discounts, calls to super to compute subtotal 
+        and restores values.
+        :param self: recordset of purchase order lines
+        :param call_super: super function object to override
+        """
+        for line in self:
+            # Keep these values for restore later
+            price = line.price_unit
+            discount = line.discount
+            line.price_unit *= (1 - (line.discount or 0.0) / 100.0)
+            line.price_unit *= (1 - (line.discount2 or 0.0) / 100.0)
+            line.price_unit *= (1 - (line.discount3 or 0.0) / 100.0)
+            # Avoid recalculate discount
+            line.discount = 0.0
+            # Call a super that uses price_unit to compute subtotal
+            call_super()
```

Comment:
Is what I want to. I need to make sure that's the object I'm sending to super. Otherwise I was having weird cache issues (fields reseting values) here: https://github.com/odoo/odoo/blob/10.0/addons/purchase/models/purchase.py#L517

Not to mention that I need to bridge purchase_discount methods so discount is not calculated twice.

Maybe I could convert this into a singleton method and loop it (or not) from the outside